### PR TITLE
Add initial charm structure

### PR DESCRIPTION
**Pull request by marceloneppel**
This PR introduces the following changes:

- README.md: Added repository description and usage instructions.
- Added the initial charm structure created by the `charmcraft init` call (with an initial PostgreSQL instance running - without replication - just to initialize the charm structure).

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    def _on_config_changed(self, _):
+        """Handle the config-changed event"""
+        # Get the postgresql container so we can configure/manipulate it
+        container = self.unit.get_container("postgresql")
+        # Create a new config layer
+        layer = self._postgresql_layer()
+
+        if container.can_connect():
+            # Get the current config
+            services = container.get_plan().to_dict().get("services", {})
```
Comment (DnPlas) [inline src/charm.py:30]: I believe you can access the `services` attribute directly with `container.get_plan().services`. See [here](https://github.com/canonical/operator/blob/aed34f9bcebf24effb6f74fab80c6c7fdfc87dda/ops/pebble.py#L433).

File: .flake8 [hunk 2]
```diff
@@ -0,0 +1,9 @@
+[flake8]
+max-line-length = 99
+select: E,W,F,C,N
+exclude:
+  venv
+  .git
+  build
+  dist
+  *.egg_info
```
Comment (jnsgruk) [inline .flake8:9]: This is probably for the next PR, but check out the structure laid out in https://github.com/canonical/operator-template for a more comprehensive layout where this stuff is moved into pyproject.toml and the repo is set up for unit/integration testing with tox.

File: README.md [hunk 3]
```diff
@@ -1 +1,47 @@
...
+$ git clone https://github.com/canonical/postgresql-k8s-operator
+$ cd postgresql-k8s-operator
+# Build the charm
+$ charmcraft pack
+# Create a model for our deployment
+$ juju add-model postgres
+# Deploy the charm (providing a password for the postgres user)
+$ juju deploy ./postgresql-k8s-operator*.charm \
+    --resource postgresql-image=ubuntu/postgres \
+    --config postgres-password=SECRET_PASSWORD
```
Comment (jnsgruk) [inline README.md:28]: In general, I'd suggest having the charm *generate* any required password like this, but again that can follow in the next PR.

File: README.md [hunk 4]
```diff
@@ -1 +1,47 @@
...
+In order to install this charm, execute the following commands (replace `SECRET_PASSWORD` with a strong password):
+
+```bash
+# Clone this charm respository
+$ git clone https://github.com/canonical/postgresql-k8s-operator
+$ cd postgresql-k8s-operator
+# Build the charm
+$ charmcraft pack
+# Create a model for our deployment
+$ juju add-model postgres
```
Comment (jnsgruk) [inline README.md:24]: Once this is published, I'd generally move the manual deployment instructions into `CONTRIBUTING.md` - see https://github.com/jnsgruk/kubernetes-dashboard-operator/blob/main/CONTRIBUTING.md

The README should be more user focused than developer focused :)

File: metadata.yaml [hunk 5]
```diff
@@ -0,0 +1,19 @@
+# Copyright 2021 Canonical
+# See LICENSE file for licensing details.
+
+name: postgresql-k8s-operator
```
Comment (jnsgruk) [inline metadata.yaml:4]:
```suggestion
name: postgresql-k8s
```

File: metadata.yaml [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+display-name: |
+  PostgreSQL K8s
```
Comment (jnsgruk) [inline metadata.yaml:6]:
```suggestion
display-name: PostgreSQL K8s
```

File: requirements-dev.txt [hunk 7]
```diff
@@ -0,0 +1,3 @@
+-r requirements.txt
+coverage
+flake8
```
Comment (jnsgruk) [inline requirements-dev.txt:3]: Again consider the structure in the operator template (but in a future PR)

File: src/charm.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+class PostgresqlOperatorCharm(CharmBase):
+    """Charmed Operator for the PostgreSQL database."""
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
```
Comment (jnsgruk) [inline src/charm.py:19]: Is there a specific reason you're not subscribing to the `pebble-ready` event?

File: src/charm.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        # Get the postgresql container so we can configure/manipulate it
+        container = self.unit.get_container("postgresql")
+        # Create a new config layer
+        layer = self._postgresql_layer()
+
+        if container.can_connect():
+            # Get the current config
+            services = container.get_plan().to_dict().get("services", {})
+            # Check if there are any changes to services
+            if services != layer["services"]:
```
Comment (DnPlas) [inline src/charm.py:32]: This is okay, though for a cleaner and more straight forward code, I'd suggest you return a `Layer` type in the `_postgresql_layer()` method below, that way you can access all attributes from it, and do something like:
 ```python
if services.services != layer.services:
...
```
See [here](https://github.com/canonical/operator/blob/aed34f9bcebf24effb6f74fab80c6c7fdfc87dda/ops/pebble.py#L456) for reference.

File: src/charm.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+                container.restart("postgresql")
+                logging.info("Restarted postgresql service")
+            # All is well, set an ActiveStatus
+            self.unit.status = ActiveStatus()
+        else:
+            self.unit.status = WaitingStatus("waiting for Pebble in workload container")
+
+    def _postgresql_layer(self):
+        """Returns a Pebble configuration layer for PostgreSQL"""
+        return {
```
Comment (jnsgruk) [inline src/charm.py:46]: You could consider returning an `ops.pebble.Layer` here

File: src/charm.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+                logging.info("Added updated layer 'postgresql' to Pebble plan")
+                # Restart it and report a new status to Juju
+                container.restart("postgresql")
+                logging.info("Restarted postgresql service")
+            # All is well, set an ActiveStatus
+            self.unit.status = ActiveStatus()
+        else:
+            self.unit.status = WaitingStatus("waiting for Pebble in workload container")
+
+    def _postgresql_layer(self):
```
Comment (jnsgruk) [inline src/charm.py:44]: Type hint for the return value?

```suggestion
    def _postgresql_layer(self) -> dict:
```

Which would be changed to ` -> Layer` if you take the suggestion below :)

**Review by jnsgruk — CHANGES_REQUESTED** [review]
Thanks Marcelo - only change I'd really like to see right now over future PRs is changing the name of the operator in `metadata.yaml`.

File: metadata.yaml (hunk 6 above)
Comment (jnsgruk) [reply metadata.yaml:6]: I'm also content for you to drop this field -- it's not really used anywhere.

**Review by DnPlas — COMMENTED** [review]
Few comments in `charm.py`, other than that, LGTM.

File: README.md (hunk 3 above)
Comment (marceloneppel) [reply README.md:28]: Thanks for the suggestion, Jon. I saw something similar in https://github.com/canonical/mongodb-operator/blob/4090146e11d8d05e5c05128a39346d3781144cae/src/charm.py#L273 and I'll implement in the next PR.

File: .flake8 (hunk 2 above)
Comment (marceloneppel) [reply .flake8:9]: Thanks Jon. I started a new PR reorganizing the repository structure to accommodate the structure in the template in order to have a more comprehensive layout for the tests.

File: src/charm.py (hunk 8 above)
Comment (marceloneppel) [reply src/charm.py:19]: There is no specific reason, Jon. As I haven't used any calls inside the pebble ready event at this time, I suppressed the function for now.

File: src/charm.py (hunk 9 above)
Comment (marceloneppel) [reply src/charm.py]: Great suggestion! Thanks for the tip Daniela! Applied in a new commit.

File: src/charm.py (hunk 1 above)
Comment (marceloneppel) [reply src/charm.py]: Thanks for the suggestion, Daniela! It makes the code better. I applied it.

File: src/charm.py (hunk 11 above)
Comment (marceloneppel) [reply src/charm.py]: Thanks for the suggestion, Jon. I applied it in a new commit using `-> Layer`.

File: src/charm.py (hunk 10 above)
Comment (marceloneppel) [reply src/charm.py]: Great. Type hint added.

File: requirements-dev.txt (hunk 7 above)
Comment (marceloneppel) [reply requirements-dev.txt:3]: I took a note of it for the next PR.

File: README.md (hunk 4 above)
Comment (marceloneppel) [reply README.md:24]: Good advice. I will move these instructions in a new PR before publishing the package.

**Comment by marceloneppel** [comment]
Closed as #2 also implements this changes and the suggestions above.

**Pull request closed without merging**